staging/src/k8s.io/kube-aggregator/pkg/controllers/openapiv3/aggregator/aggregator.go

```diff
@@ -0,0 +1,204 @@
+		if handler == nil {
+			continue
+		}
+
+		apiServiceName := fmt.Sprintf(localDelegateChainNamePattern, i)
+		localAPIService := v1.APIService{}
+		localAPIService.Name = apiServiceName
+		s.AddUpdateAPIService(handler, &localAPIService)
+		s.UpdateAPIServiceSpec(apiServiceName)
+		i++
+	}
+
+	return s, nil
+}
+
+// AddUpdateAPIService adds or updates the api service. It is thread safe.
+func (s *specAggregator) AddUpdateAPIService(handler http.Handler, apiservice *v1.APIService) {
+	s.rwMutex.Lock()
+	defer s.rwMutex.Unlock()
+	// If the APIService is being updated, use the existing struct.
+	if apiServiceInfo, ok := s.openAPIV3Specs[apiservice.Name]; ok {
+		apiServiceInfo.apiService = *apiservice
+		apiServiceInfo.handler = handler
+	}
+	s.openAPIV3Specs[apiservice.Name] = &openAPIV3APIServiceInfo{
+		apiService: *apiservice,
+		handler:    handler,
+		specs:      make(map[string]*openAPIV3SpecInfo),
+	}
+}
+
+// UpdateAPIServiceSpec updates all the OpenAPI v3 specs that the APIService serves.
+// It is thread safe.
+func (s *specAggregator) UpdateAPIServiceSpec(apiServiceName string) error {
+	s.rwMutex.Lock()
+	defer s.rwMutex.Unlock()
+
+	apiService, exists := s.openAPIV3Specs[apiServiceName]
+	if !exists {
+		return fmt.Errorf("APIService %s does not exist for update", apiServiceName)
+	}
+
+	// Pass a list of old etags to the Downloader to prevent transfers if etags match
+	etagList := make(map[string]string)
+	for gv, specInfo := range apiService.specs {
+		etagList[gv] = specInfo.etag
+	}
+	groups, err := s.downloader.Download(apiService.handler, etagList)
+	if err != nil {
+		return err
+	}
+
+	// Remove any groups that do not exist anymore
+	for group := range s.openAPIV3Specs[apiServiceName].specs {
+		if _, exists := groups[group]; !exists {
+			s.openAPIV3VersionedService.DeleteGroupVersion(group)
+			delete(s.openAPIV3Specs[apiServiceName].specs, group)
+		}
+	}
+
+	for group, info := range groups {
+		if info.spec == nil {
+			continue
+		}
+
+		// If ETag has not changed, no update is necessary
+		oldInfo, exists := s.openAPIV3Specs[apiServiceName].specs[group]
+		if exists && oldInfo.etag == info.etag {
+			continue
+		}
+		s.openAPIV3Specs[apiServiceName].specs[group] = &openAPIV3SpecInfo{
+			spec: info.spec,
+			etag: info.etag,
+		}
+		s.openAPIV3VersionedService.UpdateGroupVersion(group, info.spec)
+	}
+	return nil
+}
+
+type specAggregator struct {
+	// mutex protects all members of this struct.
+	rwMutex sync.RWMutex
+
+	// OpenAPI V3 specs by APIService and then by group/version
+	openAPIV3Specs map[string]*openAPIV3APIServiceInfo
+	// provided for dynamic OpenAPI spec
+	openAPIV3VersionedService *handler3.OpenAPIService
+
+	// For downloading the OpenAPI v3 specs from apiservices
+	downloader Downloader
+}
+
+var _ SpecAggregator = &specAggregator{}
+
+type openAPIV3APIServiceInfo struct {
+	apiService v1.APIService
+	handler    http.Handler
+	specs      map[string]*openAPIV3SpecInfo
```

Comment:
why a map? Isn't there just one per APIService?

Reply:
The apiextensions apiserver can publish multiple groups so I just adopted that kind of format here. For other apiservices, yes it's just one per.